Source/Core/Core/Debugger/BranchWatch.h

```diff
@@ -0,0 +1,286 @@
+                                                     BranchWatchSelectionInspection other)
+{
+  return self = self | other;
+}
+
+class BranchWatchCollection final
+    : public std::unordered_map<BranchWatchCollectionKey, BranchWatchCollectionValue>
+{
+};
+
+struct BranchWatchSelectionValueType
+{
+  using Inspection = BranchWatchSelectionInspection;
+
+  BranchWatchCollection::value_type* collection_ptr;
+  bool is_virtual;
+  bool condition;
+  // This is moreso a GUI thing, but it works best in the Core code for multiple reasons.
+  Inspection inspection;
+};
+
+class BranchWatchSelection final : public std::vector<BranchWatchSelectionValueType>
+{
+};
+
+enum class BranchWatchPhase : bool
+{
+  Blacklist,
+  Reduction,
+};
+
+class BranchWatch final  // Class is final to enforce the safety of GetOffsetOfRecordingActive().
+{
+public:
+  using Collection = BranchWatchCollection;
+  using Selection = BranchWatchSelection;
+  using Phase = BranchWatchPhase;
+  using SelectionInspection = BranchWatchSelectionInspection;
+
+  bool GetRecordingActive() const { return m_recording_active; }
+  void SetRecordingActive(bool active) { m_recording_active = active; }
+  void Start() { SetRecordingActive(true); }
+  void Pause() { SetRecordingActive(false); }
+  void Clear(const CPUThreadGuard& guard);
+
+  void Save(const CPUThreadGuard& guard, std::FILE* file) const;
+  void Load(const CPUThreadGuard& guard, std::FILE* file);
+
+  void IsolateHasExecuted(const CPUThreadGuard& guard);
+  void IsolateNotExecuted(const CPUThreadGuard& guard);
+  void IsolateWasOverwritten(const CPUThreadGuard& guard);
+  void IsolateNotOverwritten(const CPUThreadGuard& guard);
+  void UpdateHitsSnapshot();
+  void ClearSelectionInspection();
+  void SetSelectedInspected(std::size_t idx, SelectionInspection inspection);
+
+  Selection& GetSelection() { return m_selection; }
+  const Selection& GetSelection() const { return m_selection; }
+
+  std::size_t GetCollectionSize() const
+  {
+    return m_collection_vt.size() + m_collection_vf.size() + m_collection_pt.size() +
+           m_collection_pf.size();
+  }
+  std::size_t GetBlacklistSize() const { return m_blacklist_size; }
+  Phase GetRecordingPhase() const { return m_recording_phase; };
+
+  // An empty selection in reduction mode can't be reconstructed when loading from a file.
+  bool CanSave() const { return !(m_recording_phase == Phase::Reduction && m_selection.empty()); }
+
+  // TODO C++20: constexpr w/ std::bit_cast
+  static inline u64 FK(u32 origin, u32 destination)
+  {
+    return Common::BitCast<u64>(FakeBranchWatchCollectionKey{origin, destination});
+  }
+
+  // All Hit member functions are for the CPUThread only
+  // HitXX_fk are optimized for when origin and destination can be passed in one register easily.
+  // HitXX_fk_n are the same, but also increment the total_hits by N (see dcbx JIT code).
+  static void HitVT_fk(BranchWatch* branch_watch, u64 fake_key, u32 inst)
```

Comment:
All of these function names are not particularly self-explanatory. Can you maybe make them more clear or at least put some comments on what exactly they mean?

Also I assume they're static so that the JIT can call them easily? That should probably be in the comments here too.

Reply:
Yes, these functions are static because I am unsure if `ABI_CallFunction` is compatible with the thiscall calling convention. I made the first parameter a pointer instead of a reference for a similar reason. Though I later found some JIT code which passes pointer parameters to functions expecting references, I'm not confident that's necessarily safe.
I think I disagree that the function names need further elaboration. I already explain the meaning behind the less obvious "fk" and "n" additions to the function name in the above comment and parameter names. The VT / VF / PT / PF portion of the function name can be figured out by looking at which collection member is being modified in the body of the function (self-documenting). These members have, in turn, comments at their declarations clarifying V / P is for Virtual / Physical, and T / F is for True Path / False Path. Must that be repeated everywhere VT / VF / PT / PF gets brought up?

Reply:
Alright, I have caved and will offer two options to expand the abbreviated function names. The reason I shortened them in the first place was to keep things terse, and for that reason I prefer the commit named "Abbreviated" over the follow-up named "Full". Let me know which one is better.

Reply:
Personally I prefer the Full over the Abbreviated, it's more clear at the call site without the direct context of seeing the BranchWatch internals next to it and the functions aren't used often enough to benefit from shortening IMO.